src/analysis/normed/group/seminorm.lean

```diff
@@ -69,11 +83,87 @@ and such that `f x⁻¹ = f x` and `f x = 0 → x = 1` for all `x`. -/
 (eq_one_of_map_eq_zero' : ∀ x, to_fun x = 0 → x = 1)
 
-attribute [nolint doc_blame] add_group_seminorm.to_zero_hom  add_group_norm.to_add_group_seminorm
-  group_norm.to_group_seminorm
+/-- A nonarchimedean norm on an additive group `G` is a function `f : G → ℝ` that preserves zero, is
+nonarchimedean and such that `f (-x) = f x` and `f x = 0 → x = 0` for all `x`. -/
+@[protect_proj]
+structure nonarch_add_group_norm (G : Type*) [add_group G] extends nonarch_add_group_seminorm G :=
+(eq_zero_of_map_eq_zero' : ∀ x, to_fun x = 0 → x = 0)
+
+attribute [nolint doc_blame] add_group_seminorm.to_zero_hom add_group_norm.to_add_group_seminorm
+  group_norm.to_group_seminorm nonarch_add_group_seminorm.to_zero_hom
+  nonarch_add_group_norm.to_nonarch_add_group_seminorm
 
 attribute [to_additive] group_norm.to_group_seminorm
 
+/-- `nonarch_add_group_seminorm_class F α` states that `F` is a type of nonarchimedean seminorms on
+the additive group `α`.
+
+You should extend this class when you extend `nonarch_add_group_seminorm`. -/
+class nonarch_add_group_seminorm_class (F : Type*) (α : out_param $ Type*) [add_group α]
+  extends nonarchimedean_hom_class F α ℝ :=
+(map_zero (f : F) : f 0 = 0)
+(map_neg_eq_map' (f : F) (a : α) : f (-a) = f a)
+
+/-- `nonarch_add_group_norm_class F α` states that `F` is a type of nonarchimedean norms on the
+additive group `α`.
+
+You should extend this class when you extend `nonarch_add_group_norm`. -/
+class nonarch_add_group_norm_class (F : Type*) (α : out_param $ Type*) [add_group α]
+  extends nonarch_add_group_seminorm_class F α :=
+(eq_zero_of_map_eq_zero (f : F) {a : α} : f a = 0 → a = 0)
+
+@[priority 100] -- See note [lower instance priority]
+instance nonarch_add_group_seminorm_class.to_zero_hom_class [add_group E]
+  [nonarch_add_group_seminorm_class F E] :
+  zero_hom_class F E ℝ :=
+{ ..‹nonarch_add_group_seminorm_class F E› }
+
+section nonarch_add_group_seminorm_class
+variables [add_group E] [nonarch_add_group_seminorm_class F E] (f : F) (x y : E)
+include E
+
+lemma map_sub_le_max : f (x - y) ≤ max (f x) (f y) :=
+by { rw [sub_eq_add_neg, ← nonarch_add_group_seminorm_class.map_neg_eq_map' f y],
+     exact map_add_le_max _ _ _ }
+
+end nonarch_add_group_seminorm_class
+
+@[priority 100] -- See note [lower instance priority]
+instance nonarch_group_seminorm_class.to_nonneg_hom_class [add_group E]
+  [nonarch_add_group_seminorm_class F E] :
+  nonneg_hom_class F E ℝ :=
+{ map_nonneg := λ f a, begin
+    rw [← map_zero f, ← sub_self a],
+    exact le_trans (map_sub_le_max _ _ _) (by rw max_self (f a)),
+  end,
+  ..‹nonarch_add_group_seminorm_class F E› }
+
+section nonarch_add_group_seminorm_class
+variables [add_group E] [nonarch_add_group_seminorm_class F E] (f : F) (x y : E)
+include E
+
+lemma map_add_le_add' : f (x + y) ≤  f x + f y :=
+le_trans (map_add_le_max _ _ _)
+  (max_le (le_add_of_nonneg_right (map_nonneg _ _)) (le_add_of_nonneg_left (map_nonneg _ _)))
```

Comment:
This one at least can go inside the instance

Reply:
Should we use `[subadditive_hom_class](https://leanprover-community.github.io/mathlib_docs/algebra/order/hom/basic.html#subadditive_hom_class)` here or it's automatic because of the instances below?

Reply:
It's automatic because of the instances below.